{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4a05992c",
   "metadata": {},
   "source": [
    "# MAGS:\n",
    "1. Assembly using MEGAhit, keep everything over 2000bp (n result = 1,341,809)\n",
    "- Assembly was done in Snakemake\n",
    "- https://www.nature.com/articles/s41586-019-0965-1.pdf (2000 bp)\n",
    "\n",
    "2. Try deduplicating with CD-hit at 99% ID\n",
    "- https://journals.asm.org/doi/full/10.1128/mSystems.00245-21 (99% identity (v4.6.8, −c 0.99 −aS 0.99))\n",
    "- https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-019-0625-6 (100% ident, 100% ID)\n",
    "- https://www.pnas.org/doi/10.1073/pnas.2105124118 (99% ANI)\n",
    "\n",
    "3. mapping with Bowtie2 or bbMap\n",
    "\n",
    "a. Bowtie2:\n",
    "https://www.sciencedirect.com/science/article/pii/S0092867419300017 (option ‘--very-sensitive-local’)\n",
    "- Viral and metabolic controls on high rates of microbial sulfur and carbon cycling in wetland ecosystems\n",
    "\n",
    "b. BBmap (probs better):\n",
    "- https://www.biorxiv.org/content/10.1101/2022.05.02.490339v1.full (“fast=t ambig=random minid=0.98“)\n",
    "- https://journals.asm.org/doi/full/10.1128/mSystems.00245-21 (requiring 95% identity)\n",
    "\n",
    "\n",
    "4.  Then make depth files : \n",
    "- https://bitbucket.org/berkeleylab/metabat/wiki/Home\n",
    "\n",
    "5. Then use metabat2:\n",
    "- https://www.nature.com/articles/s41586-019-0965-1 (option --minContig 2000) and default parameters)\n",
    "- https://www.sciencedirect.com/science/article/pii/S0092867419300017 (default?)\n",
    "- https://www.nature.com/articles/s41564-018-0276-6#Sec2 (—veryspecific’ flag, Jess paper)\n",
    "\n",
    "Or metabat, concoct and maxbin, followed by DASTool\n",
    "\n",
    "Then:\n",
    "6. CheckM \n",
    "- https://www.nature.com/articles/s41586-019-0965-1 (near complete: >90% complete, <5% contam, or medium qual: >50% complete, <10% contam)\n",
    "- https://www.nature.com/articles/s41564-018-0276-6#Sec2 \n",
    "- https://www.nature.com/articles/s41587-020-0603-3#Sec11 (>50% complete, <5% contamimation)\n",
    "- https://www.sciencedirect.com/science/article/pii/S0092867419300017\n",
    "\n",
    "So: keep everyting >50% complete, <10% contam, and divide into near complete and medium quality.\n",
    "\n",
    "7. Use refineM to remove potential MAG contamination\n",
    "\n",
    "8. Use dRep to dereplicate\n",
    "- https://www.nature.com/articles/s41586-019-0965-1 \n",
    "dRep was run with options -pa 0.9 (primary cluster at 90%), -sa 0.95 (secondary cluster at 95%), -cm larger (coverage method: larger), -con 5 (contamination threshold of 5%). For the near-complete MAGs, the -nc parameter was set to 0.60 (coverage threshold of 60%), whereas for the medium-quality MAGs with a QS >50 this was changed to 0.30 (coverage threshold of 30%)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2d272e0",
   "metadata": {},
   "source": [
    "## Bbmap package for sequence wrangling\n",
    "- https://jgi.doe.gov/data-and-tools/software-tools/bbtools/\n",
    "- https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0185056"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b19448b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename, length selection, sorting with bbmap package:\n",
    "\n",
    "# Load bbmap package\n",
    "module load bbmap/38-72\n",
    "\n",
    "# rename contigs bc spaces in names from megahit\n",
    "rename.sh in=all_bulk_contigs.fa out=all_bulk_contigs_renamed.fa prefix=BodegaBay_bulk_\n",
    "\n",
    "# Keep only longer than 2000 bp\n",
    "reformat.sh in=all_bulk_contigs_renamed.fa out=all_bulk_contigs_renamed_len.fa minlength=2000\n",
    "\n",
    "# Sort by lenght\n",
    "sortbyname.sh in=all_bulk_contigs_renamed_len.fa out=all_bulk_contigs_renamed_len.sorted.fa length descending\n",
    "\n",
    "for f in *.fa\n",
    "do\n",
    "sortbyname.sh in=$f out=${f%_raw*}.sorted.fa length descending\n",
    "done\n",
    "\n",
    "\n",
    "# cut into smaller files bc cdhit not working (mem issues)\n",
    "partition.sh in=../220518_all_contigs.fa out=out_%_raw.fa ways=10"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e55ef9c6",
   "metadata": {},
   "source": [
    "## CD-hit to deduplicate\n",
    "- http://weizhong-lab.ucsd.edu/cd-hit/\n",
    "- Cd-hit: a fast program for clustering and comparing large sets of protein or nucleotide sequences\", Weizhong Li & Adam Godzik Bioinformatics, (2006) 22:1658-9\n",
    "- Limin Fu, Beifang Niu, Zhengwei Zhu, Sitao Wu and Weizhong Li, CD-HIT: accelerated for clustering the next generation sequencing data. Bioinformatics, (2012), 28 (23): 3150-3152. doi: 10.1093/bioinformatics/bts565\n",
    "- Num of resulting sequences: 1,019,242"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcd0372e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cd-hit to deduplicate the sequences\n",
    "\n",
    "\n",
    "module load cdhit/2007-013\n",
    "\n",
    "# Run cd-hit at 99% ident\n",
    "cd-hit-est -i all_bulk_contigs_renamed_sorted.fa \\\n",
    "-o all_bulk_contigs_renamed_cluster.fa \\\n",
    "-c 0.99 -aS 0.99 -M 150000 -T 24 \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae0b55f7",
   "metadata": {},
   "source": [
    "# Use bowtie2 to map reads to these contigs\n",
    "\n",
    "- Make a bowtie2 index\n",
    "- map reads to index using bowtie2\n",
    "- http://bowtie-bio.sourceforge.net/bowtie2/index.shtml\n",
    "- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3322381/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d4dd0af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build index file\n",
    "module load bowtie2.3.0\n",
    "bowtie2-build ./all_bulk_contigs_renamed_cluster.fa all_bulk_contigs_cluster --threads 12\n",
    "\n",
    "# Map the reads\n",
    "module load bowtie2.3.0\n",
    "\n",
    "bowtie2 --threads 12 --sensitive \\\n",
    "-x /home/amhorst/2022_wetlands/assemblies/coverage/all_bulk_contigs_cluster \\\n",
    "-1 $1 \\\n",
    "-2 ${1%_R1*}_R2.fastq.gz \\\n",
    "-S /home/amhorst/2022_wetlands/assemblies/coverage/mapping/${1%%_R1*}.sam \\\n",
    "--no-unal --sensitive "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd8e8def",
   "metadata": {},
   "source": [
    "# Convert samfiles to bamfiles using samtools\n",
    "- index bamfiles using samtools\n",
    "- http://www.htslib.org/\n",
    "- https://academic.oup.com/bioinformatics/article/25/16/2078/204688\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "703c5b49",
   "metadata": {},
   "outputs": [],
   "source": [
    "module load samtools/1.15.1\n",
    "\n",
    "for f in BMLS2*.sam\n",
    "do\n",
    "echo $f\n",
    "samtools view -@ 12 -F 4 -bS $f| samtools sort > ${f%.sam*}.bam \n",
    "&& \\\n",
    "samtools index ${f%_L002*}.bam\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f259668f",
   "metadata": {},
   "source": [
    "# Use metabat2 for binning\n",
    "- https://bitbucket.org/berkeleylab/metabat/src/master/\n",
    "- First make a depth file from all the .bam files\n",
    "- https://github.com/GenomicsAotearoa/metagenomics_summer_school/blob/master/materials/day2/ex7_initial_binning.md\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4728929e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# depth file for MetaBAT2\n",
    "module load metabat2/2.12.1\n",
    "jgi_summarize_bam_contig_depths --outputDepth metabat_depth.txt *.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ec99681",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create bins\n",
    "module load metabat2/2.12.1\n",
    "\n",
    "metabat -i all_bulk_contigs_renamed_cluster.fa \\\n",
    "-a ../coverage/mapping/metabat_depth.txt \n",
    "-o ../metabat/bodega_bay_metabat -t 20\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46cd12eb",
   "metadata": {},
   "source": [
    "# Use checkM to estimate completness and contamination of bins\n",
    "- https://ecogenomics.github.io/CheckM/\n",
    "- Keep everyting >50% complete, <10% contaminated\n",
    "- Divide into near complete (>90% complete, <5% contam) and medium quality (>50% complete, <10% contam). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b13bd4cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# checkM \n",
    "#!/bin/bash\n",
    "#SBATCH --job-name=checkM\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH -t 05:30:00 \n",
    "#SBATCH --ntasks=20\n",
    "#SBATCH --output=./checkM_%j.out\n",
    "#SBATCH --error=./checkM_%j.err\n",
    "#SBATCH --partition=bmm\n",
    "\n",
    "module load checkm-genome/1.0.13\n",
    "source activate checkm-genome-1.0.13\n",
    "  \n",
    "checkm lineage_wf ./bins/ checkm -x fa -t 20"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "475aad5c",
   "metadata": {},
   "source": [
    "## Use RefineM to remove contamination\n",
    "- https://github.com/dparks1134/RefineM\n",
    "- v 0.1.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d76b0c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# install refinem\n",
    "source ~/.bashrc\n",
    "conda activate refinem (v0.1.2)\n",
    "\n",
    "\n",
    "\n",
    "# Run refineM scaffold stats\n",
    ">refinem scaffold_stats -c 16 <scaffold_file> <bin_dir> <stats_output_dir> <bam_files>\n",
    "\n",
    "refinem scaffold_stats -c 10 \\\n",
    "all_bulk_contigs_renamed_cluster.fa \\\n",
    "../metabat/ \\\n",
    "../refineM/ \\\n",
    "./mapping/*.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b065773",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH --job-name=refineM\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH -t 05:30:00 \n",
    "#SBATCH --ntasks=20\n",
    "#SBATCH --output=./refineM_%j.out\n",
    "#SBATCH --error=./refineM_%j.err\n",
    "#SBATCH --partition=bmm\n",
    "\n",
    "\n",
    "source ~/.bashrc\n",
    "conda activate refinem \n",
    "\n",
    "refinem scaffold_stats -c 20 \\\n",
    "all_bulk_contigs_renamed_cluster.fa \\\n",
    "../metabat/ \\\n",
    "../refineM/ \\\n",
    "./mapping/*.bam --genome_ext fa\n",
    "\n",
    "\n",
    "#!/bin/bash\n",
    "#SBATCH --job-name=refineM\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH -t 05:30:00 \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --output=./refineM_%j.out\n",
    "#SBATCH --error=./refineM_%j.err\n",
    "#SBATCH --partition=bmm\n",
    "\n",
    "# RefineM remove outliers\n",
    "source ~/.bashrc\n",
    "conda activate refinem \n",
    "\n",
    "refinem outliers ../refineM/scaffold_stats.tsv ../refineM/outliers && \\\n",
    "refinem filter_bins ../metabat/ ../refineM/outliers/outliers.tsv ../bins_filter_refinem\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6418bfb1",
   "metadata": {},
   "source": [
    "# Dereplicate MAGs (dRep v3.2.0)\n",
    "- https://www.sciencedirect.com/science/article/pii/S0092867419300017\n",
    "-  options -pa 0.9 (primary cluster at 90%), -sa 0.95 (secondary cluster at 95%), -cm larger (coverage method: larger), -con 5 (contamination threshold of 5%). -nc 0.30 (coverage threshold of 30%)\n",
    "- Use Alexandre Almeida's code on github (take from snakemake file)\n",
    "- dRep will also run CheckM so that's nice\n",
    "- Redo again "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c42f8e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dereplicate using drep in: ./2022_wetlands/MAG_bins\n",
    "\n",
    "#!/bin/bash\n",
    "#SBATCH --job-name=drep\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH -t 16:00:00\n",
    "#SBATCH --ntasks=24\n",
    "#SBATCH --partition=bmm\n",
    "#SBATCH --output=./drep_%j.out\n",
    "#SBATCH --error=./drep_%j.err\n",
    "#SBATCH --mem=100GB\n",
    "\n",
    "source ~/.bashrc\n",
    "conda activate drep\n",
    "\n",
    "# Load other programs that dRep needs\n",
    "module load mummer\n",
    "module load mash\n",
    "module load bowtie2\n",
    "\n",
    "dRep dereplicate \\\n",
    "dRep_MAGs_output \\\n",
    "-p 16 \\\n",
    "-g ./bins_filter_refinem/*.fa \\\n",
    "-pa 0.9 -sa 0.95 -nc 0.30 -cm larger \\\n",
    "-comp 50 -con 10 -l 1000 \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa951c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dereplicate using drep in: ./2022_wetlands/MAG_bins\n",
    "\n",
    "#!/bin/bash\n",
    "#SBATCH --job-name=drep\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH -t 5:00:00\n",
    "#SBATCH --ntasks=24\n",
    "#SBATCH --partition=bmm\n",
    "#SBATCH --output=./drep_%j.out\n",
    "#SBATCH --error=./drep_%j.err\n",
    "#SBATCH --mem=50GB\n",
    "\n",
    "source ~/.bashrc\n",
    "conda activate drep\n",
    "\n",
    "# Load other programs that dRep needs\n",
    "module load mummer\n",
    "module load mash\n",
    "module load bowtie2\n",
    "\n",
    "dRep dereplicate \\\n",
    "dRep_MAGs_output \\\n",
    "-p 24 \\\n",
    "-g ./all_bins_refineM_renamed/*.fa \\\n",
    "-pa 0.9 -sa 0.95 -nc 0.30 -cm larger \\\n",
    "-comp 50 -con 10 -l 1000 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "897eb69a",
   "metadata": {},
   "source": [
    "# Use CheckM to estimate completness, contamination and get taxonomy of the bins\n",
    "- https://ecogenomics.github.io/CheckM/\n",
    "- Keep everyting >50% complete, <10% contaminated\n",
    "- Divide into near complete (>90% complete, <5% contam) and medium quality (>50% complete, <10% contam). \n",
    "- Marker lineage: indicates lineage used for inferring marker set (a precise indication of where a bin was placed in CheckM's reference tree can be obtained with the tree_qa command)\n",
    "-  The concatenated fasta file of aligned marker genes is in: ./checkM/storage/tree/concatenated.fa\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ac583d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load checkm\n",
    "\n",
    "#!/bin/bash\n",
    "#SBATCH --job-name=checkm_tree\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH -t 02:00:00 \n",
    "#SBATCH --ntasks=20\n",
    "#SBATCH --output=tree.out\n",
    "#SBATCH --error=tree.err\n",
    "#SBATCH --partition=bmm\n",
    "\n",
    "module load checkm-genome/1.0.13\n",
    "source activate checkm-genome-1.0.13\n",
    "\n",
    "# First make a tree\n",
    "checkm tree ./dereplicated_genomes/ ./checkM -x .fa -t 20\n",
    "\n",
    "# Then assess the phylogenetic markers within the tree\n",
    "checkm tree_qa ./checkM -o 2 \n",
    "\n",
    "# Create a marker set for all the bins that were first processed by tree making\n",
    "checkm lineage_set ./checkM marker_file_qual_bins\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f9bf342",
   "metadata": {},
   "source": [
    "## We ran crass a while back for finding CRISPR-Spacer pairs\n",
    "https://ctskennerton.github.io/crass/assets/manual.pdf\n",
    "- First run crass to identify spacers and repeats\n",
    "- then use blast to match these to MAGs and vOTUs\n",
    "- blast all spacers to vOTUs\n",
    "- blast repeats to magbins\n",
    "- See SPRUCE paper on the specifics and citations on why we chose those\n",
    "- v 1.0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1ee76a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# crass for crispr from raw reads\n",
    "# Do crass too for the 2019 samples\n",
    "source ~/.bashrc\n",
    "conda activate crass\n",
    "\n",
    "for f in *R1*.gz; do\n",
    "echo $f\n",
    "echo ${f%%_R1*}_R2_paired.fq.gz\n",
    "echo ${f%%_R1*}\n",
    "sbatch crass.sh $f \n",
    "done\n",
    "\n",
    "#!/bin/bash\n",
    "#SBATCH --job-name=crass\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH -t 02:30:00 \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --output=../../crass/out/crass_%j.out\n",
    "#SBATCH --error=../../crass/err/crass_%j.err\n",
    "#SBATCH --partition=bmm\n",
    "\n",
    "\n",
    "# load module\n",
    "source ~/.bashrc\n",
    "conda activate crass\n",
    "\n",
    "crass $1 ${1%%_R1*}_R2.fastq.gz -o ../../crass/${1%%_R1*} -l 4\n",
    "\n",
    "\n",
    "# extract spacer and repeat sequences\n",
    "for f in BB_*/crass.crispr; do\n",
    "echo $f\n",
    "crisprtools extract -o ${f%%crass.crispr*}spacers -s $f -x \n",
    "crisprtools extract -o ${f%%crass.crispr*}repeats -d $f -x\n",
    "done\n",
    "\n",
    "\n",
    "# all the repeats and spacers need a prefix with the sample name, cause there is multiple spacers/repeats with the same name\n",
    "module load bbmap\n",
    "for f in BB_*\n",
    "do\n",
    "echo $f\n",
    "cd $f \n",
    "# concatenate \n",
    "cat ./spacers/*.fa > all_spacers_crass.fa\n",
    "# rename\n",
    "rename.sh in=all_spacers_crass.fa out=all_spacers_crass_rename.fa prefix=$f addprefix=t\n",
    "cat ./repeats/*.fa > all_repeats_crass.fa\n",
    "rename.sh in=all_repeats_crass.fa out=all_repeats_crass_rename.fa prefix=$f addprefix=t\n",
    "cd ..\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "# concatenate all spacers\n",
    "cat */all_spacers_crass_rename.fa > all_spacers_crass.fa\n",
    "\n",
    "# concatenate all repeats\n",
    "cat */all_repeats_crass_rename.fa > all_repeats_crass.fa\n",
    "\n",
    "# deduplicate by sequence\n",
    "source ~/.bashrc\n",
    "conda activate seqkit\n",
    "\n",
    "seqkit rmdup --quiet -s all_spacers_crass.fa > all_spacers_crass.dedupe.fa\n",
    "seqkit rmdup --quiet -s all_repeats_crass.fa > all_repeats_crass.dedupe.fa\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f21f7cfa",
   "metadata": {},
   "source": [
    "## Use blast to match ( blast/2.7.1):\n",
    "- blast all spacers to vOTUs\n",
    "- blast repeats to magbins\n",
    "\n",
    "- https://www.nature.com/articles/s41396-021-01132-4#Sec9 \n",
    "- That paper only says less than 2 mismatches for spacers. \n",
    "\n",
    "- https://www.nature.com/articles/s41564-022-01150-8#Sec6\n",
    "( CRISPR spacer blastn-short alignments to a UViG (indicated by phage-like icons) are shown, with arrows representing 100% identity, 0 mismatches (solid) and 96%–99.9% identity, 1 mismatch (dashed)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "574de66e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a total blast db of all MAG seqs\n",
    "# first rename the bin sequences by putting a prefix with the bin name\n",
    "for f in *.fa \n",
    "do \n",
    "echo ${f%%.filtered.fa*}\n",
    "rename.sh in=$f out=${f%%.fa*}_rename.fa prefix=${f%%.filtered.fa*} addprefix=t\n",
    "done\n",
    "\n",
    "cat *_rename.fa >> all_qual_bins_rename.fa \n",
    "\n",
    "# Now make blastdb\n",
    "module load blast \n",
    "makeblastdb -in all_qual_bins_rename.fa -dbtype nucl -out qual_bins\n",
    "\n",
    "\n",
    "# Make a blastdb of all vOTU sequences\n",
    "makeblastdb -in 220518_drep_all_vOTUs_recov_BB.fa -dbtype nucl -out vOTU_BB_db\n",
    "\n",
    "# blast back blastn -eval 1e-10 and 100% nucl ident\n",
    "\n",
    "# WE BLAST THE SPACER SEQUENCES TO THE VOTUS\n",
    "blastn -task blastn-short \\\n",
    "-query all_spacers_crass.dedupe.fa \\\n",
    "-db ../outputs/vOTU_BB_db -evalue 1e-3 -perc_identity 95 -outfmt 6 \\\n",
    "-out spacers_vOTUs_crass_noeval.tsv\n",
    "\n",
    "# blastn REPEAT back to all MAG sequences. \n",
    "blastn -task blastn-short \\\n",
    "-query all_repeats_crass.fa \\\n",
    "-db ./220907_dRep_MAGs \\\n",
    "-evalue 1e-2 -perc_identity 95 -outfmt 6 \\\n",
    "-out repeats_prok_crass_noeval.tsv\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db647200",
   "metadata": {},
   "source": [
    "# create a phylogenetic tree\n",
    "\n",
    "Hug 2016:  A maximum likelihood tree was inferred as described for the concatenated ribosomal protein trees, under the LG plus gamma model of evolution (PROTGAMMALG in the RAxML model section), and with the number of bootstraps automatically determined (MRE-based bootstopping criterion). https://www.nature.com/articles/nmicrobiol201648!#Sec1\n",
    "\n",
    "- Fasta file made by checkM is in: Fasta contains\n",
    "- concatenated predicted protein alignment of 43 marker genes defined by CheckM \n",
    "- use RaxML https://www.nature.com/articles/s41467-019-13443-4\n",
    "- https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1009200#sec009\n",
    "- RAxML used 250 bootstraps, using the auto bootstopping\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94b33bfd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a4bdb4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# I am an idiot and to run raxml with multiple threads you need raxmlHPC-PTHREADS\n",
    "\n",
    "#!/bin/bash\n",
    "#SBATCH --job-name=raxml_pthreads\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH -t 4:00:00 \n",
    "#SBATCH --ntasks=5\n",
    "#SBATCH --output=raxml_%j.out\n",
    "#SBATCH --error=raxml_%j.err\n",
    "#SBATCH --partition=bmm\n",
    "\n",
    "source ~/.bashrc\n",
    "conda activate raxml\n",
    "\n",
    "raxmlHPC-PTHREADS \\\n",
    "-m PROTGAMMALG \\\n",
    "-x 12345 -p 12345 -# autoMRE \\\n",
    "-T 250 \\\n",
    "-n MAG_tree \\\n",
    "-s ../checkM_markergenes_concat.fa.reduced"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc0f92a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Then use bootstrap to create the best tree with bootstrap values and branch lengths.\n",
    "raxmlHPC -f b \\\n",
    "-m PROTGAMMALG \\\n",
    "-n test_bootstrap \\\n",
    "-t RAxML_bestTree* \\\n",
    "-z RAxML_bootstrap.MAG_tree_60\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb6352bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a tree only for bins with a crispr link\n",
    "# linearize the file\n",
    "#remove enters\n",
    "awk '/^>/{print s? s\"\\n\"$0:$0;s=\"\";next}{s=s sprintf(\"%s\",$0)}END{if(s)print s}' \\\n",
    "markergenes_copy.fa > markergenes_linear.faa ; done\n",
    "\n",
    "# get only protein seqs of interest from the file\n",
    "while read line; do grep -A 1 \"$line\" markergenes_linear.faa; done < bins_w_crispr_link.txt > bins_w_crispr_link.faa\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "018e33e9",
   "metadata": {},
   "source": [
    "## Mapping reads using bowtie2\n",
    "- Make index\n",
    "- Map reads using --sensitive, don't include unaligned reads bc of file size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c569c0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Index MAG fasta using Bowtie2\n",
    "module load bowtie2\n",
    "bowtie2-build ../220907_dRep_MAGs.fa \\\n",
    "--large-index  \\\n",
    "./220907_dRep_MAGs -p 16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5abf705c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map reads to this file\n",
    "bowtie2 --threads 16 --sensitive \\\n",
    "-x ./220907_dRep_MAGs \\\n",
    "-1 $1 \\\n",
    "-2 ${1%_R1*}_R2.fastq.gz \\\n",
    "-S ./${1%%_R1*}.sam \\\n",
    "--no-unal"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54ade89a",
   "metadata": {},
   "source": [
    "## Coverage table using CoverM\n",
    "- Singleton et al., https://bio-protocol.org/exchange/minidetail?type=30&id=9291876\n",
    "- Olm et al., https://www.science.org/doi/full/10.1126/science.abj2972#supplementary-materials\n",
    "\n",
    "default settings except:\n",
    "coverm genome -m relative_abundance --min-read-aligned-percent 0.75 --min-read-percent-identity 0.95 --min-covered-fraction .5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb00e9e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# EITHER\n",
    "source ~/.bashrc\n",
    "conda activate coverm\n",
    "\n",
    "coverm genome \\\n",
    "--genome-fasta-directory ../dereplicated_genomes \\\n",
    "-m trimmed_mean --min-covered-fraction 0.5 \\\n",
    " -b *.bam > 221208_coverM_bins_05_tm.csv -x fa\n",
    "\n",
    "# OR\n",
    "source ~/.bashrc\n",
    "conda activate coverm\n",
    "\n",
    "coverm genome  --genome-fasta-directory ../dereplicated_genomes \\\n",
    "-m relative_abundance --min-read-aligned-percent 0.75 \\\n",
    "--min-read-percent-identity 0.95 -x fa \\\n",
    "--min-covered-fraction 0.5 -b *.bam > 220929_coverM_bins.csv\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
